Sources/Fuzzilli/Evaluation/ProgramCoverageEvaluator.swift

```diff
@@ -47,18 +60,37 @@ public class ProgramCoverageEvaluator: ComponentBase, ProgramEvaluator {
     private var context = libcoverage.cov_context()
     
     public init(runner: ScriptRunner) {
-        super.init(name: "Coverage")
-        
+        // In order to keep clean abstractions, any corpus scheduler requiring edge counting
+        // needs to call nableEdgeTracking(), via downcasting of ProgramEvaluator
+        self.shouldTrackEdges = false
+
+        super.init(name: "Coverage")        
+
         let id = ProgramCoverageEvaluator.instances
         ProgramCoverageEvaluator.instances += 1
         
+
         context.id = Int32(id)
         guard libcoverage.cov_initialize(&context) == 0 else {
             fatalError("Could not initialize libcoverage")
         }
         runner.setEnvironmentVariable("SHM_ID", to: "shm_id_\(getpid())_\(id)")
     }
     
+    public func enableEdgeTracking() {
+        shouldTrackEdges = true
+    }
+
+    public func getEdgeCountPtr() -> UnsafeMutableBufferPointer<UInt64>? {
```

Comment:
The array currently is being mutated [here](https://github.com/googleprojectzero/fuzzilli/pull/171/files#diff-d11396d14389331fb850f0bc5f4d8ecee8aff27c291b9f114b4a47dda952ee6bR156). This is done to intentionally age all selected samples, which avoids non-deterministic ones from continually being selected.

Alternatively, I could add an age parameter in this method call, which gets passed to the coverage code, or add it as a separate function call. This function could either age all edges with hit counts <= some specified amount (not taking dropout into account), or MarkovCorpus could pass an array of its chosen edges, to be specifically incremented. What do you think?

Reply:
Ah right, I see. But even with this ageing process, wouldn't non-deterministic edges still be selected more frequently than deterministic ones? For example, if there is a deterministic edge E1 and a non-deterministic one E2 that only triggers once, and both have a count of 1, then after selecting both in the MarkovCorpus and doing N iterations, their counts should now be roughly `counts[E1] = 1 + ageIncrement + N` and `counts[E2] = 1 + ageIncrement` I think, and so the E2 (non-deterministic) will be selected again (I guess in reality, it would now first select another edge E3, but that shouldn't matter too much).

Would it be an option to instead re-execute every sample that is added to the corpus and to remove every edge from `edgeMap` which is not discovered again in the 2nd run? I guess the edge counts array could be used to compute non-deterministic edges this way, or there could be a new API in the coverage evaluator to temporarily reset some edges or something like that.

Reply:
I agree that non-deterministic will still be selected more. My view here is that intentionally aging prevents the case where (ignoring dropout) the corpus selects the same N non-deterministic samples each round, while the aging allows edge E3 to execute, in your example.

What are you thoughts on adding two API functions, one to query the edge count at a given index, and another to clear both the edge count and virgin bit at a given index? The MarkovCorpus would run each sample a second time (maybe have this value be configurable?), and use these two functions to check for non-determinism, and potentially clear the edges found by a non-deterministic sample. This would replace the current aging functionality. WDYT?